src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -106,6 +106,36 @@ protected override void ReleaseResources()
             }
         }
 
+        protected override Task InvokeResultAsync(IActionResult result)
+        {
+            // We also have some special initialization we need to do for PageResult.
+            if (result is PageResult pageResult)
+            {
+                // If we used a PageModel then the Page isn't initialized yet.
+                if (_viewContext == null)
+                {
+                    _viewContext = new ViewContext(
+                        _pageContext,
+                        NullView.Instance,
+                        _pageContext.ViewData,
+                        _tempDataFactory.GetTempData(_pageContext.HttpContext),
+                        TextWriter.Null,
+                        _htmlHelperOptions);
+                    _viewContext.ExecutingFilePath = _pageContext.ActionDescriptor.RelativePath;
+                }
+
+                if (_page == null)
+                {
+                    _page = (PageBase)CacheEntry.PageFactory(_pageContext, _viewContext);
+                }
+
+                pageResult.Page = _page;
+                pageResult.ViewData = pageResult.ViewData ?? _pageContext.ViewData;
+            }
+
+            return base.InvokeResultAsync(result);
+        }
```

Comment:
This is a breaking change. Previously filters could observe that `pageResult.Page` was intialized.

Reply:
We might decide that it's OK based on the fact that views work similarly, and based on the fact that it fixes a bug.

Reply:
> This is a breaking change. Previously filters could observe that pageResult.Page was intialized.

Good point. Should we do this in both places in that case? i.e after the pagehandler method is executed and even in InvokeResultAsync? We can extract out this logic into a common method and call from both these places.

Reply:
I'm thinking about it...... Doing that will introduce another bug where the PageFactory is called to create a page, but then not called to dispose it.

Reply:
I think the scenario of writing a filter that modifies the Page is a bit broken no matter what.

Reply:
I'm leaning towards your original fix.

Reply:
> I think the scenario of writing a filter that modifies the Page is a bit broken no matter what.

We don't have any coverage for manipulating the page in a filter, it would seem like it's not a reasonable scenario.